### PR TITLE
Implement tag auto-completion by history

### PR DESCRIPTION
**Pull request by akihikodaki**
This is a functionality similar to one implemented in Pawoo:
https://github.com/pixiv/mastodon/commit/21a3c70f8083b1347d2b8420ed7001b78c2c9620

UPDATE: @unarist pointed out this is an alternative for #3196.

### CODE REVIEW

**Comment by abcang** [comment]
I think that complementing by `/api/v1/search` is more convenient than complementing by history. Why is it compatible with Pawoo implementation?

**Comment by akihikodaki** [comment]
> I think that complementing by /api/v1/search is more convenient than complementing by history. Why is it compatible with Pawoo implementation?

To be honest I do not always appreciate account suggestion much because it suggests lots of people but the number of people I often mention is quite a limited.

A tag will be tooted even more often than mentions for an account. Suggestions ordered by recent uses would be a better choice.

File: app/javascript/mastodon/actions/compose.js [hunk 1]
```diff
@@ -322,6 +340,48 @@ export function selectComposeSuggestion(position, token, suggestion) {
...
+export function updateTagHistory(tags) {
+  return {
+    type: COMPOSE_TAG_HISTORY_UPDATE,
+    tags,
+  };
+}
+
+export function hydrateCompose() {
+  return (dispatch, getState) => {
+    const me = getState().getIn(['meta', 'me']);
```
Comment (abcang) [inline app/javascript/mastodon/actions/compose.js]: `me` can be acquired by importing initial_state.

`import { me } from '../initial_state';`

**Comment by abcang** [comment]
I understood. I agree broadly.

Complementing by using api can choose the correct tag. So, I thought it would be nice to implement it as another feature.

**Comment by akihikodaki** [comment]
> Complementing by using api can choose the correct tag. So, I thought it would be nice to implement it as another feature.

What do you suggest to implement as another feature? To search tags on the instance, to use history, or to combine them?

**Comment by abcang** [comment]
I believe it is best to combine tag search and history. But, I accepts this PR even without implements of tag search.

**Review by Gargron — APPROVED** [review]

**Pull request merged**